src/xenia/hid/sdlpad/sdlpad_input_driver.cc

```diff
@@ -0,0 +1,402 @@
+/**
+ ******************************************************************************
+ * Xenia : Xbox 360 Emulator Research Project                                 *
+ ******************************************************************************
+ * Copyright 2019 Ben Vanik. All rights reserved.                             *
+ * Released under the BSD license - see LICENSE in the root for more details. *
+ ******************************************************************************
+ */
+
+#include "xenia/hid/sdlpad/sdlpad_input_driver.h"
+
+#if XE_PLATFORM_WIN32
+#include "xenia/base/platform_win.h"
+#endif  // XE_PLATFORM_WIN32
+
+#include "xenia/base/cvar.h"
+#include "xenia/ui/window.h"
+
+// TODO(joellinn) make this path relative to the config folder.
+DEFINE_string(mappings_file, "gamecontrollerdb.txt",
+              "Filename of a database with custom game controller mappings.",
+              "SDL");
+
+namespace xe {
+namespace hid {
+namespace sdlpad {
+
+SDLPadInputDriver::SDLPadInputDriver(xe::ui::Window* window)
+    : InputDriver(window),
+      sdl_events_initialized_(false),
+      sdl_gamecontroller_initialized_(false),
+      sdl_events_unflushed_(0),
+      controllers_(),
+      controllers_mutex_() {}
+
+SDLPadInputDriver::~SDLPadInputDriver() {
+  for (size_t i = 0; i < controllers_.size(); i++) {
+    if (controllers_.at(i).sdl) {
+      SDL_GameControllerClose(controllers_.at(i).sdl);
+      controllers_.at(i) = {};
+    }
+  }
+  if (sdl_events_initialized_) {
+    SDL_QuitSubSystem(SDL_INIT_EVENTS);
+    sdl_events_initialized_ = false;
+  }
+  if (sdl_gamecontroller_initialized_) {
+    SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+    sdl_gamecontroller_initialized_ = false;
+  }
+}
+
+X_STATUS SDLPadInputDriver::Setup() {
+  if (!TestSDLVersion()) {
+    return X_STATUS_UNSUCCESSFUL;
+  }
+
+  // SDL_PumpEvents should only be run in the thread that initialized SDL - we
+  // are hijacking the window loop thread for that.
+  window_->loop()->PostSynchronous([&]() {
```

Comment:
This actually requires linking with `xenia-ui`. This should probably be added to the premake file.

Reply:
mh, ok that link is missing on xinuput too